at_lookup/lib/src/cache/secondary_address_cache.dart

```diff
@@ -0,0 +1,120 @@
+
+  final String _rootDomain;
+  final int _rootPort;
+  late SecondaryAddressFinder _secondaryFinder;
+
+  SecondaryAddressCache(this._rootDomain, this._rootPort, {SecondaryAddressFinder? secondaryFinder}) {
+    _secondaryFinder = secondaryFinder ?? SecondaryAddressFinder();
+  }
+
+  bool cacheContains(String atSign) {
+    return _map.containsKey(stripAtSignFromAtSign(atSign));
+  }
+
+  /// Returns the expiry time of this entry in millisecondsSinceEpoch, or null if this atSign is not in cache
+  int? getCacheExpiryTime(String atSign) {
+    atSign = stripAtSignFromAtSign(atSign);
+    if (_map.containsKey(atSign)) {
+      return _map[atSign]!.expiresAt;
+    } else {
+      return null;
+    }
+  }
+
+  Future<SecondaryAddress> getAddress(String atSign, {bool refreshCacheNow = false, Duration cacheFor = defaultCacheDuration}) async {
+    atSign = stripAtSignFromAtSign(atSign);
+
+    if (refreshCacheNow || _cacheIsEmptyOrExpired(atSign)) {
+      // _updateCache will either populate the cache, or throw an exception
+      await _updateCache(atSign, cacheFor);
+    }
+
+    if (_map.containsKey(atSign)) {
+      // should always be true, since _updateCache will throw an exception if it fails
+      return _map[atSign]!.secondaryAddress;
+    } else {
+      // but just in case, we'll throw an exception if it's not
+      throw Exception("Failed to find secondary, in a theoretically impossible way");
+    }
+  }
+
+  bool _cacheIsEmptyOrExpired(String atSign) {
+    if (_map.containsKey(atSign)) {
+      SecondaryAddressCacheEntry entry = _map[atSign]!;
+      if (entry.expiresAt < DateTime.now().millisecondsSinceEpoch) {
+        // expiresAt is in the past - cache has expired
+        return true;
+      } else {
+        // cache has not yet expired
+        return false;
+      }
+    } else {
+      // cache is empty
+      return true;
+    }
+  }
+
+  static String stripAtSignFromAtSign(String atSign) {
+    if (atSign.startsWith('@')) {
+      atSign = atSign.replaceFirst('@', '');
+    }
+    return atSign;
+  }
+  static String getNotFoundExceptionMessage(String atSign) {
+    return 'Unable to find secondary address for atSign:$atSign';
+  }
+  Future<void> _updateCache(String atSign, Duration cacheFor) async {
+    try {
+      String? secondaryUrl = await _secondaryFinder.findSecondary(atSign, _rootDomain, _rootPort);
+      if (secondaryUrl == null ||
+          secondaryUrl.isEmpty ||
+          secondaryUrl == 'data:null') {
+        throw SecondaryNotFoundException(getNotFoundExceptionMessage(atSign));
+      }
+      var secondaryInfo = LookUpUtil.getSecondaryInfo(secondaryUrl);
+      String secondaryHost = secondaryInfo[0];
+      int secondaryPort = int.parse(secondaryInfo[1]);
+
+      SecondaryAddress addr = SecondaryAddress(secondaryHost, secondaryPort);
+      _map[atSign] = SecondaryAddressCacheEntry(addr, DateTime.now().add(cacheFor).millisecondsSinceEpoch);
+    } on Exception catch (e) {
```

Comment:
Have been wondering more about this. One of the problems we have is when we throw an exception. So I think we should just return the cached value (even if it is stale, it's better than throwing an exception) - and perhaps for visibility, add another field to the CacheEntry like 'lookupException' which would be 'most recent cache refresh fail reason'; would be set when we catch an exception, would be cleared when a lookup succeeds.